os/fs/smartfs/smartfs_utils.c

```diff
@@ -1065,209 +1163,276 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 			offset += entrysize;
 			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 		}
+		ret_entry->dsector = psector;
+		psector = nextsector;
+	}
 
-		/* If we found an entry, stop the search */
-
-		if (found) {
-			break;
-		}
-
-		/* If there are no more sectors, then we need to add one to make
-		 * room for the new entry.
-		 */
-
-		if (nextsector == SMARTFS_ERASEDSTATE_16BIT) {
-			/* Allocate a new sector and chain it to the last one */
-
-			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-			if (ret < 0) {
-				goto errout;
-			}
+errout:
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	if (ret != OK) {
+		ret = -ENOENT;
+	}
 
-			nextsector = (uint16_t)ret;
+	if (ret == OK) {
+		RESET_ENTRY_FLAGS(ret_entry);
+	}
 
-			/* Chain the next sector into this sector sector */
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	fdbg("\nfIND_FIRSTINVALIDENTRY returns invalidentry->flags = %x\n", ret_entry->flags);
+	return ret;
+}
 
-			chainheader->nextsector[0] = (uint8_t)(nextsector & 0x00FF);
-			chainheader->nextsector[1] = (uint8_t)(nextsector >> 8);
-			//*((uint16_t *)chainheader->nextsector) = nextsector;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, nextsector);
-			readwrite.count = sizeof(uint16_t);
-			readwrite.buffer = chainheader->nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error chaining sector %d\n", nextsector);
-				goto errout;
-			}
+/****************************************************************************
+ * Name: smartfs_writeentry
+ *
+ * Description: Takes care of the operation of renaming an existing file
+ *              in a particular directory to a new name in the same dircetory
+ *              while minimizing the need to allocate new sectors for
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
 
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-			 /* TODO We will modify whole of create entry to integrity MTD operation soon.
-			  * We will always create sector first, and then update inside of sector. 
-			  * This will help to clear by fs_recover because all sector will become to isolated sector...
-			  * It is temp FIX, If we use sector buffer to support CRC, currently created new parent sector which is chained 
-			  * to previous parent sector is not committed yet. So Now we break loop to prevent read error 
-			  */
-			offset = sizeof(struct smartfs_chain_header_s);
-			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
-			psector = nextsector;
-			break;
-#endif
-		}
+int smartfs_writeentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s new_entry, const char *filename, uint16_t type, mode_t mode, struct smartfs_entry_s *ret_entry)
```

Comment:
Please align callee & caller.
you always use new_entry for last parameter, but here it is ret_entry.
It is needed when some kind of type casting, but we do not cast or nothing related to casting. (I like newentry, actually =D)

Reply:
`newentry` is always a pointer to the entry that will hold the details of the newly found (in case of find_availableentry) or created (in case of createentry) entry returned for writing.

`ret_entry` is always a pointer to the entry that will hold details of the entry that was written to MTD by writeentry

So it becomes kind of self-explanatory
`newentry `- newly allocated entry for writing
`ret_entry `- returning final entry that was written to MTD

Reply:
Ohkay, I can keep the entry returned by writeentry after writing to MTD as direntry.
Let's do that.